src/ore/src/btree_map.rs

```diff
@@ -0,0 +1,180 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License in the LICENSE file at the
+// root of this repository, or online at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! BTreeMap utilities.
+
+use std::collections::btree_map::{BTreeMap, Iter};
+
+/// Extension methods for [`BTreeMap`].
+pub trait BTreeMapExt<'a, K: PartialOrd, V1, V2> {
+    /// Performs a merge join between two [`BTreeMap`]s, returning an struct that can iterate over
+    /// the join producing `(K, V1, V2)`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::collections::BTreeMap;
+    ///
+    /// use mz_ore::str::BTreeMapExt;
+    ///
+    /// let mut lhs = BTreeMap::new();
+    /// lhs.insert(1, 2);
+    /// lhs.insert(3, 4);
+    ///
+    /// let mut rhs = BTreeMap::new();
+    /// lhs.insert(3, "four");
+    /// lhs.insert(5, "six");
+    ///
+    /// let results = lhs.merge_join(&rhs).collect();
+    ///
+    /// assert_eq!(results, vec![&3, &4, "four"]);
+    /// ```
+    fn merge_join(&'a self, r: &'a BTreeMap<K, V2>) -> MergeJoin<'a, K, V1, V2>;
```

Comment:
Did you consider [`itertools::merge_join_by`](https://docs.rs/itertools/latest/itertools/fn.merge_join_by.html)?